### PR TITLE
DOC: sparse.linalg: updated preconditioner doc for iterative linear solvers

### LINKED ISSUE #5818: Clarify definition of preconditioner for sparse linear system solvers
http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.sparse.linalg.cg.html (and other solvers too) currently defines the preconditioner M as "The preconditioner should approximate the inverse of A."
[Wikipedia](https://en.wikipedia.org/wiki/Conjugate_gradient_method#Convergence_properties_of_the_conjugate_gradient_method) and [MATLAB](http://www.mathworks.com/help/matlab/ref/pcg.html) both seem to use the inverse definition (i.e. they solve `M^-1.A.x=M^-1.y`).
Both definitions seem to exist in the literature(?) but (assuming that the note is correct) I think it would be helpful to clarify which M is meant there ("Note: the definition of the preconditioner matches the definition in reference XYZ, and is the inverse of the definition in reference XYZ'.")

### PR DESCRIPTION
**Pull request by Afleloup**
Closes #5818 

#### Reference issue

    Closes #5818
#### What does this implement/fix?

    The argument M in iterative linear solver
    'bicg', 'bicgstab', 'cg', 'cgs' is not standard notation
    in comparison to the literature. It is the inverse of the
    preconditioner usually denoted M^-1. Their should be a
    precondition on M for the preconditioned conjugate gradient.
    It is now explicited in the documentation and linked
    to relevant references.
##### References

    .. [1] TREFETHEN, Lloyd N. et BAU, David. Numerical linear algebra. 
           Society for Industrial and Applied Mathematics, 2022.
    .. [2] SAAD, Yousef. Iterative methods for sparse linear systems. 
           Society for Industrial and Applied Mathematics, 2003.
    .. [3] "Conjugate Gradient Method, Wikipedia, 
           https://en.wikipedia.org/wiki/Conjugate_gradient_method
    .. [4] "Preconditioner", 
           Wikipedia, https://en.wikipedia.org/wiki/Preconditioner
    .. [5] KATRUTSA, Alexandr, BOTCHEV, Mike, OVCHINNIKOV, George, et al. 
           How to optimize preconditioners for the conjugate gradient method: 
           a stochastic approach. arXiv preprint arXiv:1806.06045, 2018.
    .. [6] CARABA, Elena. Preconditioned conjugate gradient algorithm. 2008.
    .. [7] "Biconjugate gradient stabilized method", 
           Wikipedia, https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method
    .. [8] "Conjugate gradient squared", Wikipedia,
           https://en.wikipedia.org/wiki/Conjugate_gradient_squared_method
    .. [9] "Biconjugate gradient method", Wikipedia, 
           https://en.wikipedia.org/wiki/Biconjugate_gradient_method

### CODE REVIEW

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 1]
```diff
@@ -359,6 +391,27 @@ def cg(A, b, x0=None, *, tol=_NoValue, maxiter=None, M=None, callback=None,
             0  : successful exit
             >0 : convergence to tolerance not achieved, number of iterations
 
+    Notes
+    -----
+    The inverse precondtitioner noted ``M`` is not standard notation in the litterature. It is usually denoted ``M^-1``. See [1,2,3,4]
```
Comment (thalassemia) [inline scipy/sparse/linalg/_isolve/iterative.py:396]:
```suggestion
    The inverse preconditioner noted ``M`` is not standard notation in the literature. It is usually denoted ``M^-1``. See [1,2,3,4]
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 2]
```diff
@@ -474,6 +527,22 @@ def cgs(A, b, x0=None, *, tol=_NoValue, maxiter=None, M=None, callback=None,
             >0 : convergence to tolerance not achieved, number of iterations
             <0 : parameter breakdown
 
+    Notes
+    -----
+    The inverse precondititioner noted ``M`` is not standard notation in the litterature. It is usually denoted ``M^-1``. See [1,2,3]
```
Comment (thalassemia) [inline scipy/sparse/linalg/_isolve/iterative.py:532]:
```suggestion
    The inverse preconditioner noted ``M`` is not standard notation in the literature. It is usually denoted ``M^-1``. See [1,2,3]
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 3]
```diff
@@ -215,6 +230,22 @@ def bicgstab(A, b, *, x0=None, tol=_NoValue, maxiter=None, M=None,
             0  : successful exit
             >0 : convergence to tolerance not achieved, number of iterations
             <0 : parameter breakdown
+    
+    Notes
+    -----
+    The inverse precondtitioner noted ``M`` is not standard notation in the litterature. It is usually denoted ``M^-1``. See [1,2,3]
```
Comment (thalassemia) [inline scipy/sparse/linalg/_isolve/iterative.py:236]:
```suggestion
    The inverse preconditioner noted ``M`` is not standard notation in the literature. It is usually denoted ``M^-1``. See [1,2,3]
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 4]
```diff
@@ -215,6 +230,22 @@ def bicgstab(A, b, *, x0=None, tol=_NoValue, maxiter=None, M=None,
...
+    for comparison.
+    Here, ``M`` should be a matrix such that MA has a smaller condition number than ``A``.
```
Comment (thalassemia) [inline scipy/sparse/linalg/_isolve/iterative.py:238]:
```suggestion
    Here, ``M`` should be a matrix such that ``MA`` has a smaller condition number than ``A``.
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 5]
```diff
@@ -87,6 +88,21 @@ def bicg(A, b, x0=None, *, tol=_NoValue, maxiter=None, M=None, callback=None,
             >0 : convergence to tolerance not achieved, number of iterations
             <0 : parameter breakdown
 
+    Notes
+    -----
+    The inverse precondtitioner noted ``M`` is not standard notation in the litterature. It is usually denoted ``M^-1``. See [1,2,3]
```
Comment (thalassemia) [inline scipy/sparse/linalg/_isolve/iterative.py:93]:
```suggestion
    The inverse preconditioner noted ``M`` is not standard notation in the literature. It is usually denoted ``M^-1``. See [1,2,3]
```

**Review by thalassemia — CHANGES_REQUESTED** [review]

**Comment by thalassemia** [comment]
These are just typo fixes. Someone familiar with the math still needs to review this.

**Comment by ilayn** [comment]
Thank you @Afleloup 

I think this is conflating the point here a bit too much. Just because the argument name is `M` is not meant to be that we find a preconditioner and invert it and then pass here. We can correct the docs saying simply $MA$ should improve the conditioning of the problem as an extra sentence, and that should be enough.

The theoretical treatment is irrelevant here since the operator name can be anything. It's utility is important. 

The point of all the inverses attached to M is to emphasize that it kinda sorta inverts A. But in fact it does a more important thing which is, ideally, improves the condition number of the LHS. Hence inverse is a red herring here. The existing docs read correct to me, but can be clarified for the mismatch with the literature. 

Note that the preconditioner definition is all over the place, as noted with the wiki page you added as a reference

>  It is also common to call $T=P^{-1}$ the preconditioner, rather than $\displaystyle P$, since $P$ itself is rarely explicitly available.

**Comment by Afleloup** [comment]
Thanks, I will change it this way then. I didn't see the sentence you mentioned in the wikipedia article and thought the literature I had was consistent with the other notation.

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 6]
```diff
@@ -61,6 +60,18 @@ def bicg(A, b, x0=None, *, rtol=1e-5, atol=0., maxiter=None, M=None, callback=No
             >0 : convergence to tolerance not achieved, number of iterations
             <0 : parameter breakdown
 
+    Notes
+    -----
+    The preconditioner `M` should be a matrix such that ``M @ A`` has a smaller
+    condition number than `A`. See [1]_ for more info.
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:66]:
```suggestion
    condition number than `A`, see [1]_ .
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 7]
```diff
@@ -184,6 +192,18 @@ def bicgstab(A, b, x0=None, *, rtol=1e-5, atol=0., maxiter=None, M=None,
             >0 : convergence to tolerance not achieved, number of iterations
             <0 : parameter breakdown
 
+    Notes
+    -----
+    The preconditioner `M` should be a matrix such that ``M @ A`` has a smaller
+    condition number than `A`, see [1]_ for more info.
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:198]:
```suggestion
    condition number than `A`, see [1]_ .
```

**Review by j-bowhay — APPROVED** [review]
Thanks @Afleloup! I made a few edits to get this merged

**Pull request merged**